### PR TITLE
Annotation to disable cudagraphs for a graph

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #175271
* __->__ #174103

  [inductor] Add override_cudagraphs decorator to selectively enable/disable cudagraphs              
                                                 
  Adds a decorator API that marks functions/methods so that any compiled graph segment inlining them skips cudagraph recording. Annotation is per-graph — other segments after a graph break
  are unaffected.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

### CODE REVIEW

File: torch/_inductor/__init__.py [hunk 1]
```diff
@@ -444,3 +444,13 @@ def standalone_compile(
     return standalone_compile(
         gm, example_inputs, dynamic_shapes=dynamic_shapes, options=options, aot=aot
     )
+
+
+import dataclasses
```
Comment (BoyuanFeng) [inline torch/_inductor/__init__.py]: do we want to move import to the top of file?

File: torch/_inductor/compile_fx.py [hunk 2]
```diff
@@ -2413,6 +2432,12 @@ def compile_fx_backward(
             model_outputs_node.meta["user_visible_output_idxs"] = []
 
         fixed = count_tangents(gm)
+
+        # Check if cudagraphs should be disabled for backward via annotation
+        cudagraphs = compiler_config_extra.cudagraphs
+        if compiler_config_extra.disable_cudagraphs_for_bwd:
```
Comment (BoyuanFeng) [inline torch/_inductor/compile_fx.py]: iiuc, `compiler_config_extra.cudagraphs` controls cg for fwd;  `compiler_config_extra.cudagraphs + compiler_config_extra.disable_cudagraphs_for_bwd` controls cg for bwd.

Should we be explicit and have  `compiler_config_extra.disable_cudagraphs_for_fwd` & `compiler_config_extra.disable_cudagraphs_for_bwd` seperately? Or maybe at least have the logic at the same location (e.g., in create_compiler_config_extra).

**Review by BoyuanFeng — APPROVED** [review]
Thanks! a few minor suggestions.

File: test/inductor/test_cudagraph_trees.py [hunk 3]
```diff
@@ -3057,6 +3057,208 @@ def fn_eager(x, marked_static_y):
...
+            inp = torch.randn(4, 4, device="cuda", requires_grad=True)
+            fn_compiled = torch.compile(fn, mode="reduce-overhead")
+            out = fn_compiled(inp)
+            out.sum().backward()
+
+            self.assertEqual(out, inp.detach().pow(2).sin())
+            self.assertIsNone(self.get_manager())
+
+        def test_cudagraph_annotation_disable_bwd_only(self):
+            @torch._dynamo.disable_cudagraphs(fwd=False, bwd=True)
```
Comment (desertfire) [inline test/inductor/test_cudagraph_trees.py]: I understand this PR is only providing a disable API, but my thinking is for some of the internal models, we want cudagraphs default to off and only enable it for a few graphs. To avoid future API BC issues, I think we should rename the API to enable_cudagraphs with the boolean flags meaning flipped, even if the actual enable part of implementation will come later?

File: test/inductor/test_cudagraph_trees.py [hunk 4]
```diff
@@ -3057,6 +3057,208 @@ def fn_eager(x, marked_static_y):
...
+            self.assertEqual(out, ((inp + 1) * 2).sin())
+            self.assertIsNone(self.get_manager())
+
+        def test_cudagraph_annotation_context_manager(self):
+            """Context manager disables cudagraphs for all graphs within block,
+            including across graph breaks."""
+
+            def model(x):
+                y = x + 1  # g1: outside context, cudagraphs ON
+                torch._dynamo.graph_break()
```
Comment (desertfire) [inline test/inductor/test_cudagraph_trees.py:3215]: What happens if we don't have the graph break here? I expect `y = x + 1` and `z = y * 2` will be put in one graph and controlled by the same annotation?

Comment (eellison) [reply test/inductor/test_cudagraph_trees.py:3215]: yes, that's correct, and there are other nested, tests showing this

File: test/inductor/test_cudagraph_trees.py (hunk 3 above)
Comment (eellison) [reply test/inductor/test_cudagraph_trees.py]: Hmm, i'm open to bike shedding.

`torch._dynamo.override_cudagraphs` or  `torch._dynamo.configure_cudagraphs`

`with fwd=optional[bool], bwd=optional[bool]`.

E.g. with `torch._dynamo.override_cudagraphs(fwd=False, bwd=True)`

or `torch._dynamo.override_cudagraphs(fwd=False, bwd=None)`

something like that.

Comment (desertfire) [reply test/inductor/test_cudagraph_trees.py]: `override_cudagraphs` sounds good to me. I suppose `None` means leave it as the default?

**Review by desertfire — APPROVED** [review]

File: test/inductor/test_cudagraph_trees.py (hunk 3 above)
Comment (eellison) [reply test/inductor/test_cudagraph_trees.py]: yup, None means default. i will update it .

**Review by williamwen42 — COMMENTED** [review]
dynamo changes lgtm

File: torch/_inductor/compile_fx.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/compile_fx.py]: The logic normalizes the disables with the `cudagraph` global setting, to maximize cache hits. I don't think we should have `compiler_config_extra.disable_cudagraphs_for_fwd` because it's redundant with cudagraphs being off.

File: test/inductor/test_cudagraph_trees.py [hunk 5]
```diff
@@ -3054,6 +3054,305 @@ def fn_eager(x, marked_static_y):
...
+                out.sum().backward()
+
+            self.assertEqual(out, inp.detach().pow(2).sin())
+            self.assertIsNotNone(self.get_manager())
+            # One forward and one backward
+            self.assertEqual(self.get_manager().new_graph_id().id, 2)
+
+        @torch._inductor.config.patch("triton.cudagraphs", False)
+        def test_cudagraph_annotation_enable_fwd_only(self):
+            """override_cudagraphs(fwd=True) force-enables cudagraphs for
```
Comment (Microve) [inline test/inductor/test_cudagraph_trees.py]: If I explicitly set bwd=False, will backward still be enabled?
